benchmarks/src/jmh/java/io/grpc/benchmarks/netty/HandlerRegistryBenchmark.java

```diff
@@ -0,0 +1,127 @@
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Level;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.infra.Blackhole;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Benchmark for {@link io.grpc.MutableHandlerRegistryImpl}.
+ */
+@State(Scope.Benchmark)
+@Fork(1)
+public class HandlerRegistryBenchmark {
+
+    private static final String VALID_CHARACTERS = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789.";
+
+    @Param({"50"})
+    public int nameLength;
+
+    @Param ({"100"})
+    public int serviceCount;
+
+    @Param({"100"})
+    public int methodCountPerService;
+
+    @Param({"false", "true"})
+    public boolean useNewRegistry;
+
+    private HandlerRegistry registry;
+    private OldRegistry oldRegistry;
+    private MutableHandlerRegistryImpl newRegistry;
+    private List<String> fullMethodNames;
+
+    @Setup(Level.Trial)
+    public void setup() throws Exception {
+        oldRegistry = new OldRegistry();
+        newRegistry = new MutableHandlerRegistryImpl();
+        fullMethodNames = new ArrayList<String>(serviceCount * methodCountPerService);
+        for (int serviceIndex = 0; serviceIndex < serviceCount; ++serviceIndex) {
+            String serviceName = randomString();
+            ServerServiceDefinition.Builder serviceBuilder = ServerServiceDefinition.builder(serviceName);
+            for (int methodIndex = 0; methodIndex < methodCountPerService; ++methodIndex) {
+                String methodName = randomString();
+                serviceBuilder.addMethod(ServerMethodDefinition.create(methodName, null, null, null));
+                fullMethodNames.add("/" + serviceName + "/" + methodName);
+            }
+            if (useNewRegistry) {
+                newRegistry.addService(serviceBuilder.build());
+            } else {
+                oldRegistry.addService(serviceBuilder.build());
+            }
+        }
+        registry = useNewRegistry ? newRegistry : oldRegistry;
+    }
+
+    @Benchmark
+    public void lookupMethod(Blackhole bh) {
+        for (String fullMethodName : fullMethodNames) {
+            bh.consume(registry.lookupMethod(fullMethodName));
+        }
+    }
+
+    private String randomString() {
+        Random r = new Random();
+        byte[] bytes = new byte[nameLength];
+        for (int ix = 0; ix<nameLength; ++ix) {
+            int charIx = r.nextInt(VALID_CHARACTERS.length());
+            bytes[ix] = (byte) VALID_CHARACTERS.charAt(charIx);
+        }
+        return new String(bytes);
+    }
+
+    private final class OldRegistry extends MutableHandlerRegistry {
```

Comment:
Can we not add the old registry, and instead just run the benchmark before and after the change?